### ISSUE TITLE
[@typescript-eslint/no-shadow] false positives with conditional types and `infer`

### ISSUE THREAD

**Issue by dinofx** [typescript-eslint/typescript-eslint#2526]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-shadow": ["error"]
  }
}
```

```TS
export type ArrayInput<Func> = Func extends (arg0: Array<infer T>) => any
  ? T[] // if true
  : Func extends (...args: infer T) => any // else
    ? T
    : never;

function f1(...s: string[]) {
  return s.length;
}

function f2(numbers: number[]) {
  return numbers.length;
}

export let r1: ArrayInput<typeof f1>;
export let r2: ArrayInput<typeof f2>;
```

**Expected Result**

No shadowing should be reported.  The first `infer T` is only in scope in the first "if true" expression.  In the "else" expression, it is valid to use `infer T` again, as the earlier `T` is not in scope.

**Actual Result**

A false positive is reported:

> "'T' is already declared in the upper scope."

**Versions**

This was reproduced in both 4.0.2 and 4.1.0.

**Comment by dinofx** [comment]
I tried this out briefly and it seems to help:

```TS
  protected TSConditionalType(node: TSESTree.TSConditionalType): void {
    // conditional types can define inferred type parameters
    // which are only accessible from inside the conditional parameter
    this.#referencer.scopeManager.nestConditionalTypeScope(node);

    this.visitChildren(node, ['falseType']);

    this.#referencer.close(node);

    this.visit(node.falseType);
  }
```

**Comment by bradzacher** [comment]
This bug is due to my misunderstanding of where inferred types are usable

Your fix is likely all that is needed to solve this. That plus the snapshots will need updating.

Please raise a PR! You've done the investigation so I'd love for you to have your name on the commit :)

**Comment by dinofx** [comment]
> plus the snapshots will need updating

Not sure what's needed there.  I opened a PR and included a new "valid" test case for no-shadow

**Comment by dinofx** [comment]
@bradzacher PR linked.
